File: backend/attendance_mailer.py
import pandas as pd
from datetime import datetime
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import re

def send_email(to_email, student_name, absent_dates, session_name):
    current_year = datetime.now().year
    formatted_dates = ", ".join([f"{date}-{current_year}" for date in absent_dates])

    subject = "Absence Notification"
    body = f"""
    Dear {student_name},

    We noticed that you were absent for the enrolled course {session_name} on the following days:
    {formatted_dates}

    Regular attendance is essential to stay aligned with the course content and placement activities. Please ensure you go through the missed session before attending the upcoming ones.

    Kindly note 85% attendance is mandatory to get certification and placement assistance. A minimum of 70% attendance is mandatory to be eligible for certification and placement support.

    Warm Regards,
    Learning Coordinator
    ChipEdge Technologies Pvt Ltd
    https://chipedge.com/
    """

    msg = MIMEMultipart("alternative")
    msg['Subject'] = subject
    msg['From'] = "customer.success@chipedge.com"  # replace with your sender email
    msg['To'] = to_email
    msg.attach(MIMEText(body, "plain"))

    # Use actual email credentials / environment variables for security
    with smtplib.SMTP('smtp.gmail.com', 587) as server:
        server.starttls()
        server.login("customer.success@chipedge.com", "hvxdizbuidwsitpg")  # Secure app password
        server.send_message(msg)

def process_attendance(file_path):
    # Read CSV with possible header skip and metadata lines according to your notebook
    with open(file_path, 'r') as f:
        lines = f.readlines()

    # Extract session name from line 4 (index 3)
    session_name = ""
    if len(lines) > 3:
        line_text = " ".join(lines[3].strip().split(","))
        match = re.search(r'Course\s*Name\s*[:-]*\s*(.*)', line_text, re.IGNORECASE)
        session_name = match.group(1).strip() if match else line_text

    # Load attendance data skipping first 6 rows (metadata)
    df = pd.read_csv(file_path, skiprows=6)
    df.columns = df.columns.str.strip()

    required_columns = ['Learners', 'Email']
    for col in required_columns:
        if col not in df.columns:
            return f"Missing required column: {col}"

    dates = df.columns[4:]  # Adjust if needed

    absent_students = {}
    for _, row in df.iterrows():
        student_name = row['Learners']
        email = row['Email']
        absent_dates = []
        for date in dates: # type: ignore
            cell = str(row[date]).strip().upper()
            if cell in ['A', 'OL']:
                absent_dates.append(date.split('.')[0])
        if absent_dates:
            absent_students[email] = (student_name, absent_dates)

    # Send emails
    count = 0
    for email, (student_name, absent_dates) in absent_students.items():
        send_email(email, student_name, absent_dates, session_name)
        count += 1

    return f"✅ Emails sent to {count} learner(s)."

File: backend/attendanceMailer.js
import fs from "fs";
import csvParser from "csv-parser";
import nodemailer from "nodemailer";
import dayjs from "dayjs";

// Setup your transporter with credentials from environment
const transporter = nodemailer.createTransport({
  host: "smtp.gmail.com",
  port: 587,
  secure: false,
  auth: {
    user: process.env.ATTENDANCE_EMAIL_USER,
    pass: process.env.ATTENDANCE_EMAIL_PASS,
  },
});

function isValidLearner(row) {
  // Checks for valid email and learner name, ignores NA or blank
  return (
    !!row["Email"] &&
    !!row["Learners"] &&
    /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(row["Email"]) &&
    row["Learners"].trim().toUpperCase() !== "NA"
  );
}

async function sendEmail(toEmail, studentName, absentDates, sessionName) {
  const datesStr = absentDates.join(", ");
  await transporter.sendMail({
    from: `"ChipEdge" <${process.env.EMAIL_USER}>`,
    to: toEmail,
    subject: "Absence Notification",
    html: `
      <p>Dear ${studentName},</p>
      <p>You were absent for your enrolled course <b>${sessionName}</b> on: ${datesStr}</p>
      <p>Please maintain attendance for certification and placement eligibility.</p>
      <p>Regards,<br/>ChipEdge Team</p>
    `,
  });
}

/**
 * Parses the attendance CSV and sends mails to all absent learners.
 * @param {string} filePath path to the uploaded CSV
 * @returns {Promise<string>} Message such as "Emails sent to X learners"
 */
async function processAttendanceFile(filePath) {
  return new Promise((resolve, reject) => {
    const absentees = {};
    const rows = [];
    let dateCols = [];

    fs.createReadStream(filePath)
      .pipe(csvParser())
      .on("headers", (headers) => {
        // Pick columns that match date ("dd-Jul") and exclude Notes columns
        dateCols = headers.filter(
          (col) =>
            col &&
            /^\d{2}-Jul$/.test(col.trim()) && // columns like "21-Jul", "22-Jul"
            !/Notes/i.test(col)
        );
      })
      .on("data", (row) => {
        // Only consideration: valid learners and emails
        if (isValidLearner(row)) {
          rows.push(row);
        }
      })
      .on("end", async () => {
        try {
          for (const row of rows) {
            const absentDates = [];
            for (const dateCol of dateCols) {
              const val = String(row[dateCol] || "").trim().toUpperCase();
              // Mark as absent for A or OL
              if (val === "A" || val === "OL") {
                absentDates.push(dateCol);
              }
            }
            if (absentDates.length > 0) {
              absentees[row["Email"]] = {
                name: row["Learners"],
                absentDates,
              };
            }
          }
          let sentCount = 0;
          for (const [email, { name, absentDates }] of Object.entries(absentees)) {
            await sendEmail(email, name, absentDates, "Attendance Session");
            sentCount++;
          }
          resolve(`Emails sent to ${sentCount} learners`);
        } catch (error) {
          reject(error);
        }
      })
      .on("error", reject);
  });
}

export { processAttendanceFile };
